xrootd/client/client.go

```
@@ -0,0 +1,141 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package client // import "go-hep.org/x/hep/xrootd/client"
+
+import (
+	"context"
+	"encoding/binary"
+	"io"
+	"net"
+
+	"go-hep.org/x/hep/xrootd/encoder"
+	"go-hep.org/x/hep/xrootd/mux"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+// A Client to xrootd server which allows to send requests and receive responses.
+// Concurrent requests are supported.
+// Zero value is invalid, Client should be instantiated using NewClient.
+type Client struct {
+	conn            net.Conn
+	mux             *mux.Mux
+	protocolVersion int32
+}
+
+const responseHeaderSize = 2 + 2 + 4
+
+type responseHeader struct {
+	StreamID   protocol.StreamID
+	Status     uint16
+	DataLength int32
+}
+
+// NewClient creates a client to xrootd server at address.
+// ctx defines a lifetime of the client, once it is done response handling is stopped.
+func NewClient(ctx context.Context, address string) (*Client, error) {
+	conn, err := net.Dial("tcp", address)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{conn, mux.New(), 0}
+
+	go client.consume(ctx)
+
+	if err := client.handshake(ctx); err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func (client *Client) consume(ctx context.Context) {
+	var header = &responseHeader{}
+	var headerBytes = make([]byte, responseHeaderSize)
+
+	for {
+		if _, err := io.ReadFull(client.conn, headerBytes); err != nil {
+			// TODO: handle EOF by redirection as specified at http://xrootd.org/doc/dev45/XRdv310.pdf, page 11
+		}
+
+		if err := encoder.Unmarshal(headerBytes, header); err != nil {
+			// TODO: should redirect in case if is not possible to decode a header as well?
```

here as well? add a panic here? so we don't swallow errors?